rockchip: add Radxa CM3 IO support

maurerr opened the pull request:
Hardware
--------
RockChip RK3566 ARM64 (4 cores)
- up to 8GB LPDDR4X
- 1x HDMI,
- 2x MIPI DSI
- 2x MIPI CSI2
- 1x eDP
- 1x PCIe card
- 2x SATA
- 2x USB 2.0 Host
- 1x USB 3.0
- 1x USB 2.0 OTG
- 10/100/1000 Base-T
- microSD slot
- 40-pin GPIO expansion header
- 12V DC

Radxa CM3 needs to mount on top of this IO board in order to create
complete Radxa CM3 IO board platform.

Installation
------------
Uncompress the OpenWrt sysupgrade and write it to a micro SD card or
internal eMMC using dd.

Signed-off-by: Marius Durbaca <mariusd84@gmail.com>

maurerr commented:
I believe I need some support/review from @mj22226 or @1715173329 as you have more experience adding support for rockchip. Thanks !

1715173329 commented:
Thanks, will check this tomorrow.
BTW have you tested this? It would be nice to post some bootlog here.

maurerr commented:
failing at bootloader stage :(
```
# Copy the idbloader and the u-boot image to the image at sector 0x40 and 0x4000
dd if="/data/openwrt-gh-maurerr/staging_dir/target-aarch64_generic_musl/image"/radxa-cm3-io-3566-idbloader.img of="/data/openwrt-gh-maurerr/build_dir/target-aarch64_generic_musl/linux-rockchip_armv8/tmp/openwrt-rockchip-armv8-radxa_cm3-io-ext4-sysupgrade.img.gz" seek=64 conv=notrunc
dd: failed to open '/data/openwrt-gh-maurerr/staging_dir/target-aarch64_generic_musl/image/radxa-cm3-io-3566-idbloader.img': No such file or directory
make[5]: *** [Makefile:62: /data/openwrt-gh-maurerr/build_dir/target-aarch64_generic_musl/linux-rockchip_armv8/tmp/openwrt-rockchip-armv8-radxa_cm3-io-ext4-sysupgrade.img.gz] Error 1
```

1715173329 commented on package/boot/uboot-rockchip/Makefile:
```diff
@@ -118,6 +118,22 @@ define U-Boot/rockpro64-rk3399
+  TPL:=rk3566_ddr_1056MHz_v1.18.bin
+endef
+
+define U-Boot/radxa-cm3-io
```
this should be `radxa-cm3-io-rk3566`, which matches the defconfig name.

maurerr replied:
fixed

1715173329 commented on package/boot/uboot-rockchip/Makefile:
```diff
@@ -153,6 +169,7 @@ UBOOT_TARGETS := \
   roc-cc-rk3328 \
   rock64-rk3328 \
   rock-pi-e-rk3328 \
+  cm3-io-rk3566 \
```
same here

maurerr replied:
fixed

1715173329 commented on target/linux/rockchip/image/armv8.mk:
```diff
@@ -77,6 +77,14 @@ define Device/pine64_rockpro64
+  DEVICE_VENDOR := Radxa
+  DEVICE_MODEL := CM3 IO
+  SOC := rk3566
+  DEVICE_DTS := rockchip/rk3566-radxa-cm3-io
```
This board does not fit the default UBOOT_DEVICE_NAME value, so need to declare it here manually.
`UBOOT_DEVICE_NAME := radxa-cm3-io-rk3566`

maurerr replied:
fixed

1715173329 commented on target/linux/rockchip/patches-6.1/018-v6.3-arm64-dts-rockchip-Add-rk3566-based-Radxa-Compute-Module-3.patch:
```diff
@@ -0,0 +1,386 @@
++	leds {
++		compatible = "gpio-leds";
++
++		led-0 {
```
https://github.com/openwrt/openwrt/blob/ced3fbcda18e200315cb077f7e2b362ae52ee065/target/linux/rockchip/patches-6.1/109-nanopc-t4-add-led-aliases.patch#L7-L12

it would be nice to add OpenWrt's LED aliases for showing system status (put it into a separated  patch).

maurerr replied:
addressed hopefully

maurerr commented:
building is fixed - full log  http://sprunge.us/XcrHiA

1715173329 commented:
Please rebase your PR upon the main branch, burn the image into a TF card / eMMC and catch boot logs.

maurerr commented:
here is the dmesg https://pastebin.com/Nd5cQHzR - as for console log it seems my cheap pl2303 clone doesn't supports 1.5mbps so I have to change the console to 115200 in the dts and post that too - hopefully is fine

1715173329 commented:
Kernel log looks great!

Final step is splitting your changes:
- one commit for rkbin: pack rk3566 atf/tpl blob
- one commit for rockchip uboot: add cm3 io support
- the last one commit for rockchip target: add cm3 io support

maurerr commented:
> Kernel log looks great!
> 
> Final step is splitting your changes:
> 
> * one commit for rkbin: pack rk3566 atf/tpl blob
> * one commit for rockchip uboot: add cm3 io support
> * the last one commit for rockchip target: add cm3 io support

done - no rkbin changea after all as I found rk3568 atf working just fine

1715173329 commented:
> done - no rkbin changes after all as I found rk3568 atf working just fine

tpl is required anyway.

1715173329 commented on package/boot/uboot-rockchip/Makefile:
```diff
@@ -118,6 +118,22 @@ define U-Boot/rockpro64-rk3399
+
+define U-Boot/rk3566/Default
+  BUILD_SUBTARGET:=armv8
+  DEPENDS:=+PACKAGE_u-boot-$(1):trusted-firmware-a-rk3568
```
```suggestion
  DEPENDS:=+PACKAGE_u-boot-$(1):trusted-firmware-a-rk3566
```

maurerr replied:
fixed

1715173329 commented on package/boot/rkbin/Makefile:
```diff
@@ -25,13 +25,20 @@ define Trusted-Firmware-A/Default
 endef
 
 TFA_TARGETS:= \
+    rk3566 \
```
```suggestion
	rk3566 \
```
use `<tab>` insead of 4 `<space>`

maurerr replied:
fixed

1715173329 reviewed (commented):
and, please sort (reverse) your commits
the first one for rkbin
second one for uboot
last one for target

maurerr commented:
> and, please sort (reverse) your commits the first one for rkbin second one for uboot last one for target

hopefully fixed

1715173329 commented on target/linux/rockchip/patches-6.1/111-radxa-cm3-io-add-led-aliases.patch:
```diff
@@ -0,0 +1,13 @@
+ 	aliases {
+ 		ethernet0 = &gmac1;
+ 		mmc1 = &sdmmc0;
++		led-boot = &led-1;
```
you should create a label for the led node:
```patch
--- a/arch/arm64/boot/dts/rockchip/rk3566-radxa-cm3.dtsi
+++ b/arch/arm64/boot/dts/rockchip/rk3566-radxa-cm3.dtsi
@@ -17,7 +17,7 @@
 	leds {
 		compatible = "gpio-leds";
 
-		led-0 {
+		status_led: led-0 {
 			gpios = <&gpio0 RK_PA6 GPIO_ACTIVE_HIGH>;
 			color = <LED_COLOR_ID_GREEN>;
 			function = LED_FUNCTION_STATUS;
```

then refer it:

```suggestion
+		led-boot = &status_led;
```

maurerr replied:
done... I hope :)

1715173329 commented:
Cc: @hauke

jpiccari commented:
Excited for this. The OrangePi CM4 also uses the rk3566, will this PR add support for all rk3566 based boards? And if not can anyone direct me to documentation on board specific changes?

maurerr commented:
> Excited for this. The OrangePi CM4 also uses the rk3566, will this PR add support for all rk3566 based boards? And if not can anyone direct me to documentation on board specific changes?

definitely not ALL rk3566 based boards - although some like opi cm4 might work - at least from what I briefly observed for their gmac1 (integrated 1gig interface) but please do test and let us all know if it :)

jpiccari commented:
```
...
[    0.678690] arm-scmi firmware:scmi: Failed. SCMI protocol 21 not active.
[    0.683454] arm-scmi firmware:scmi: Failed. SCMI protocol 19 not active.
[    0.685357] sdhci: Secure Digital Host Controller Interface driver
[    0.685929] sdhci: Copyright(c) Pierre Ossman
[    0.686325] Synopsys Designware Multimedia Card Interface Driver
[    0.687241] sdhci-pltfm: SDHCI platform and OF driver helper
[    0.688617] dwmmc_rockchip fe2c0000.mmc: IDMAC supports 32-bit address mode.
[    0.688949] dwmmc_rockchip fe2b0000.mmc: IDMAC supports 32-bit▒DDR Version V1.10 20210810
ln
ddrconfig:7
LPDDR4X, 324MHz
BW=32 Col=10 Bk=8 CS0 Row=16 CS1 Row=16 CS=2 Die BW=16 Size=4096MB
tdqss: cs0 dqs0: -96ps, dqs1: -241ps, dqs2: -96ps, dqs3: -168ps,
tdqss: cs1 dqs0: -96ps, dqs1: -241ps, dqs2: -96ps, dqs3: -168ps,

change to: 324MHz
PHY drv:clk:36,ca:36,DQ:29,odt:60
vrefinner:16%, vrefout:41%
dram drv:40,odt:0
clk skew:0x66

change to: 528MHz
PHY drv:clk:36,ca:36,DQ:29,odt:60
vrefinner:16%, vrefout:41%
dram drv:40,odt:0
clk skew:0x58

change to: 780MHz
PHY drv:clk:36,ca:36,DQ:29,odt:60
vrefinner:16%, vrefout:41%
dram drv:40,odt:0
clk skew:0x58

change to: 1056MHz(final freq)
PHY drv:clk:36,ca:36,DQ:29,odt:60
vrefinner:16%, vrefout:22%
dram drv:40,odt:80
vref_ca:00000071
clk skew:0x4c
cs 0:
the read training result:
DQS0:0x3c, DQS1:0x36, DQS2:0x3e, DQS3:0x3f,
min  : 0xc  0xd 0x10  0xe  0x1  0x3  0x9  0x9 , 0x7  0x9  0x3  0x1  0xc  0xd  0xd  0x9 ,
       0x9  0x1  0x9  0x9  0x7  0x4  0xa  0x5 , 0x8  0x6  0x6  0x1  0xb  0xe  0xd  0xb ,
mid  :0x28 0x29 0x2b 0x29 0x1e 0x20 0x26 0x25 ,0x24 0x26 0x20 0x1d 0x29 0x2a 0x2b 0x26 ,
      0x28 0x20 0x26 0x28 0x25 0x22 0x27 0x23 ,0x25 0x24 0x23 0x1f 0x29 0x2b 0x2b 0x28 ,
max  :0x44 0x45 0x47 0x45 0x3b 0x3d 0x43 0x41 ,0x42 0x44 0x3d 0x3a 0x47 0x47 0x49 0x43 ,
      0x48 0x40 0x44 0x47 0x43 0x41 0x45 0x42 ,0x43 0x42 0x41 0x3d 0x47 0x48 0x49 0x45 ,
range:0x38 0x38 0x37 0x37 0x3a 0x3a 0x3a 0x38 ,0x3b 0x3b 0x3a 0x39 0x3b 0x3a 0x3c 0x3a ,
      0x3f 0x3f 0x3b 0x3e 0x3c 0x3d 0x3b 0x3d ,0x3b 0x3c 0x3b 0x3c 0x3c 0x3a 0x3c 0x3a ,
the write training result:
DQS0:0x40, DQS1:0x2c, DQS2:0x40, DQS3:0x36,
min  :0x59 0x5c 0x5e 0x5d 0x53 0x53 0x59 0x59 0x5c ,0x49 0x4c 0x45 0x45 0x4f 0x51 0x50 0x4d 0x4d ,
      0x57 0x51 0x57 0x58 0x55 0x53 0x57 0x55 0x55 ,0x4a 0x49 0x4a 0x48 0x4e 0x51 0x4f 0x4f 0x4a ,
mid  :0x76 0x79 0x7a 0x79 0x6f 0x70 0x75 0x75 0x77 ,0x65 0x68 0x60 0x5f 0x6b 0x6d 0x6b 0x69 0x69 ,
      0x74 0x6e 0x74 0x75 0x72 0x70 0x73 0x72 0x72 ,0x67 0x65 0x66 0x64 0x6b 0x6e 0x6c 0x6c 0x67 ,
max  :0x94 0x96 0x97 0x96 0x8c 0x8d 0x92 0x92 0x92 ,0x81 0x85 0x7c 0x7a 0x88 0x89 0x87 0x86 0x85 ,
      0x92 0x8c 0x91 0x92 0x90 0x8d 0x8f 0x8f 0x90 ,0x85 0x82 0x83 0x81 0x88 0x8c 0x89 0x89 0x84 ,
range:0x3b 0x3a 0x39 0x39 0x39 0x3a 0x39 0x39 0x36 ,0x38 0x39 0x37 0x35 0x39 0x38 0x37 0x39 0x38 ,
      0x3b 0x3b 0x3a 0x3a 0x3b 0x3a 0x38 0x3a 0x3b ,0x3b 0x39 0x39 0x39 0x3a 0x3b 0x3a 0x3a 0x3a ,
cs 1:
the read training result:
DQS0:0x3f, DQS1:0x39, DQS2:0x3d, DQS3:0x3f,
min  : 0xc  0xc  0xf  0xe  0x2  0x3  0x8  0x8 , 0x7  0x9  0x3  0x1  0xd  0xe  0xe  0x9 ,
       0xb  0x1  0x9  0x9  0x8  0x5  0xb  0x5 , 0x8  0x6  0x7  0x1  0xd 0x10  0xe  0xb ,
mid  :0x29 0x2a 0x2c 0x2b 0x20 0x21 0x27 0x26 ,0x26 0x28 0x22 0x1f 0x2b 0x2c 0x2d 0x28 ,
      0x28 0x1f 0x25 0x27 0x25 0x22 0x27 0x22 ,0x25 0x23 0x23 0x1f 0x2a 0x2c 0x2a 0x27 ,
max  :0x47 0x48 0x4a 0x49 0x3e 0x40 0x46 0x44 ,0x45 0x47 0x41 0x3d 0x4a 0x4b 0x4d 0x47 ,
      0x45 0x3d 0x42 0x45 0x42 0x40 0x44 0x40 ,0x42 0x40 0x3f 0x3d 0x47 0x48 0x47 0x44 ,
range:0x3b 0x3c 0x3b 0x3b 0x3c 0x3d 0x3e 0x3c ,0x3e 0x3e 0x3e 0x3c 0x3d 0x3d 0x3f 0x3e ,
      0x3a 0x3c 0x39 0x3c 0x3a 0x3b 0x39 0x3b ,0x3a 0x3a 0x38 0x3c 0x3a 0x38 0x39 0x39 ,
the write training result:
DQS0:0x40, DQS1:0x2c, DQS2:0x40, DQS3:0x36,
min  :0x59 0x5c 0x5e 0x5d 0x53 0x52 0x59 0x59 0x5b ,0x48 0x4c 0x45 0x44 0x4e 0x50 0x4f 0x4d 0x4d ,
      0x57 0x53 0x57 0x58 0x57 0x55 0x57 0x56 0x56 ,0x4a 0x49 0x4a 0x48 0x4f 0x51 0x4f 0x4f 0x4a ,
mid  :0x76 0x78 0x7a 0x79 0x6f 0x6f 0x75 0x75 0x76 ,0x64 0x68 0x60 0x5f 0x6b 0x6c 0x6a 0x69 0x69 ,
      0x74 0x6f 0x74 0x74 0x73 0x71 0x72 0x71 0x72 ,0x68 0x65 0x66 0x64 0x6b 0x6e 0x6c 0x6c 0x67 ,
max  :0x93 0x95 0x96 0x95 0x8b 0x8c 0x92 0x92 0x92 ,0x81 0x85 0x7b 0x7a 0x88 0x89 0x86 0x85 0x86 ,
      0x92 0x8c 0x91 0x91 0x8f 0x8d 0x8e 0x8d 0x8e ,0x86 0x82 0x83 0x81 0x88 0x8b 0x89 0x89 0x84 ,
range:0x3a 0x39 0x38 0x38 0x38 0x3a 0x39 0x39 0x37 ,0x39 0x39 0x36 0x36 0x3a 0x39 0x37 0x38 0x39 ,
      0x3b 0x39 0x3a 0x39 0x38 0x38 0x37 0x37 0x38 ,0x3c 0x39 0x39 0x39 0x39 0x3a 0x3a 0x3a 0x3a ,
CA Training result:
cs:0 min  :0x45 0x40 0x3f 0x3b 0x3d 0x39 0x42 ,0x47 0x40 0x40 0x39 0x40 0x37 0x42 ,
cs:0 mid  :0x80 0x83 0x7a 0x7d 0x78 0x7b 0x6f ,0x82 0x82 0x7c 0x7b 0x7a 0x7a 0x6e ,
cs:0 max  :0xbc 0xc6 0xb6 0xbf 0xb4 0xbd 0x9c ,0xbd 0xc4 0xb8 0xbe 0xb5 0xbd 0x9b ,
cs:0 range:0x77 0x86 0x77 0x84 0x77 0x84 0x5a ,0x76 0x84 0x78 0x85 0x75 0x86 0x59 ,
cs:1 min  :0x43 0x43 0x3a 0x3c 0x3b 0x39 0x3d ,0x45 0x40 0x3d 0x3c 0x3d 0x39 0x41 ,
cs:1 mid  :0x81 0x83 0x79 0x7c 0x79 0x7a 0x6c ,0x82 0x81 0x7b 0x7c 0x79 0x78 0x6d ,
cs:1 max  :0xbf 0xc4 0xb9 0xbc 0xb8 0xbb 0x9b ,0xbf 0xc2 0xb9 0xbc 0xb6 0xb8 0x99 ,
cs:1 range:0x7c 0x81 0x7f 0x80 0x7d 0x82 0x5e ,0x7a 0x82 0x7c 0x80 0x79 0x7f 0x58 ,
out
U-Boot SPL board init
U-Boot SPL 2017.09-orangepi (Dec 28 2023 - 13:35:16)
unrecognized JEDEC id bytes: ff, ff, ff
...
```

Doesn't work. Seems like the kernel resets (maybe this is a panic?) and just loops back to u-boot over and over. I think the boards might use different eMMC chips and so maybe need some small tweaks. But it is encouraging to see the kernel booting right out of the box with this change. I'll see if I can get a PR working for this board but my day job is on software at a very different abstraction level so any pointers would be great.

1715173329 commented:
This is for the Radxa CM3 IO board, if you need support for other boards, please open a new PR or issue, don't send irrelevant comments here.

jpiccari commented:
> This is for the Radxa CM3 IO board, if you need support for other boards, please open a new PR or issue, don't send irrelevant comments here.

Radxa CM3 IO uses the same SoC as the orangepi CM4 and a dozen other rpi CM4 clone boards. I thought it would be relevant to confirm the possibility of compatibility and report back results so that either an issue could be opened or documentation on support be updated. It sounds like you would prefer even investigatory comments be left in some other issue.

1715173329 commented:
So? The same SoC doesn't mean the same hardware design.

maurerr commented:
> > This is for the Radxa CM3 IO board, if you need support for other boards, please open a new PR or issue, don't send irrelevant comments here.
> 
> Radxa CM3 IO uses the same SoC as the orangepi CM4 and a dozen other rpi CM4 clone boards. I thought it would be relevant to confirm the possibility of compatibility and report back results so that either an issue could be opened or documentation on support be updated. It sounds like you would prefer even investigatory comments be left in some other issue.

please let's discuss this topic further on openwrt forum

1715173329 commented:
Cc: @PolynomialDivision

PolynomialDivision commented:
Please write a complete patch header for your led alias patch 111... like done here:
- https://github.com/openwrt/openwrt/blob/main/target/linux/rockchip/patches-6.1/110-arm64-dts-rockchip-Update-LED-properties-for-NanoPi-.patch
- https://github.com/openwrt/openwrt/blob/main/target/linux/rockchip/patches-6.1/100-rockchip-use-system-LED-for-OpenWrt.patch

Although the patch will never be upstreamed, it simplifies patch maintenance.

PolynomialDivision commented:
Please also improve the rkbin commit message (write a complete setence). Thanks for your work. :)

maurerr committed:
rkbin: add rk3566 atf/tpl blobs

Currently there are no atf/tpl blobs for rk3566 SoCs
so this commit adds the prebuilt firmware from the vendor.

Signed-off-by: Marius Durbaca <mariusd84@gmail.com>

maurerr committed:
uboot-rockchip: add Radxa CM3 IO board support

Add support for the Radxa CM3 IO board.

Reviewed-by: Tianling Shen <cnsztl@immortalwrt.org>
Signed-off-by: Marius Durbaca <mariusd84@gmail.com>

maurerr committed:
rockchip: add Radxa CM3 IO board support

Hardware
--------
RockChip RK3566 ARM64 (4 cores)
- up to 8GB LPDDR4X
- 1x HDMI,
- 2x MIPI DSI
- 2x MIPI CSI2
- 1x eDP
- 1x PCIe card
- 2x SATA
- 2x USB 2.0 Host
- 1x USB 3.0
- 1x USB 2.0 OTG
- 10/100/1000 Base-T
- microSD slot
- 40-pin GPIO expansion header
- 12V DC

Radxa CM3 needs to mount on top of this IO board in order to create
complete Radxa CM3 IO board platform.

Installation
------------
Uncompress the OpenWrt sysupgrade and write it to a micro SD card or
internal eMMC using dd.

Reviewed-by: Tianling Shen <cnsztl@immortalwrt.org>
Signed-off-by: Marius Durbaca <mariusd84@gmail.com>

maurerr commented:
> The title of second commit ([3627eba](https://github.com/openwrt/openwrt/commit/3627ebac21114be4f3ee5af5d490824cd0625e32)) does not seem alright.
> 
> It should be: `uboot-rockchip: add Radxa CM3 IO board support`
> 
> For the commit message, as always, simply write: `Add support for the Radxa CM3 IO board.`
> 
> Otherwise: `Reviewed-by: Tianling Shen <cnsztl@immortalwrt.org>`

addressed

1715173329 commented:
> addressed

Sorry if I made any confusing, the third commit message (fda6d92f6c1c) looks pretty good and you don't need to change it, please keep it as-is.
This should be the last change, thank you!

maurerr commented:
> > addressed
> 
> Sorry if I made any confusing, the third commit message ([fda6d92](https://github.com/openwrt/openwrt/commit/fda6d92f6c1c60bd00180c5429461309474d4557)) looks pretty good and you don't need to change it, please keep it as-is. This should be the last change, thank you!

added back the content of the previous commit message. thanks again !
